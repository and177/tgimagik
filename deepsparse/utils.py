from dataclasses import dataclass
from queue import Queue
from typing import List, Optional

@dataclass
class Request:
    id: int
    prompt: str
    max_generated_tokens: int

@dataclass
class Batch:
    id: int
    requests: List[Request]

@dataclass
class CachedBatch:
    batch_id: int
    request_ids: List[int]

    def __len__(self):
        return len(self.request_ids)

@dataclass
class Generation:
    request_id: int
    token: Optional[str]
    token_id: Optional[str]
    stopped: bool

@dataclass
class GenerateRequest:
    prompt: str
    max_generated_tokens: int
    response_stream: Queue[Generation]