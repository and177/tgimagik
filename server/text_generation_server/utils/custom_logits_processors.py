from typing import Dict, Iterable
from transformers import PreTrainedTokenizerBase, LogitsWarper
import abc


class CustomLogitsWarperFactory(abc.ABC):
    def __init__(self, name):
        self.name = name

    @abc.abstractmethod
    def create_warper(self, tokenizer: PreTrainedTokenizerBase, params: Iterable[str]) -> LogitsWarper:
        raise NotImplementedError


class CustomLogitsProcessorsManager:
    processors: Dict[str, CustomLogitsWarperFactory] = {}
    
    @classmethod
    def register_factory(cls, factory: CustomLogitsWarperFactory):
        """Register a factory for a custom warper. This should be called by library developers in the __init__.py of their custom warper module,
        and the module should be included in the custom_modules command line argument of the text-generation-server CLI."""
        cls.processors[factory.name] = factory

    @classmethod
    def create_warper(cls, name: str, tokenizer: PreTrainedTokenizerBase, params: Iterable[str]) -> LogitsWarper:
        """Create a custom warper by name."""
        if name not in cls.processors:
            raise ValueError(f"Unknown warper {name}. Known warpers: {', '.join(cls.processors.keys())}")
        return cls.processors[name].create_warper(tokenizer, params)
