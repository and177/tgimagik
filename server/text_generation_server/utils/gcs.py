import os
from google.cloud import storage

GCS_PREFIX = "gs://"   
GCS_LOCAL_DIR = "/tmp/gcs_model"

def download_gcs_dir_to_local(gcs_dir: str, local_dir: str):
    if os.path.isdir(local_dir):
        return
    # gs://bucket_name/dir
    bucket_name = gcs_dir.split('/')[2]
    prefix = gcs_dir[len(GCS_PREFIX + bucket_name) :].strip('/') + '/'
    client = storage.Client()
    blobs = client.list_blobs(bucket_name, prefix=prefix)
    if not blobs:
        raise ValueError(f"No blobs found in {gcs_dir}")
    for blob in blobs:
        if blob.name[-1] == '/':
            continue
        file_path = blob.name[len(prefix) :].strip('/')
        local_file_path = os.path.join(local_dir, file_path)
        os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
        logger.info(f"==> Download {gcs_dir}/{file_path} to {local_file_path}.")
        blob.download_to_filename(local_file_path)
    logger.info("Download finished.")
