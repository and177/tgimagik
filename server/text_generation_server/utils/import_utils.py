import torch


def is_npu_available():
    try:
        import torch_npu  # noqa: F401
    except ImportError:
        return False

    return hasattr(torch, "npu") and torch.npu.is_available()


IS_ROCM_SYSTEM = torch.version.hip is not None
IS_CUDA_SYSTEM = torch.version.cuda is not None
IS_NPU_SYSTEM = is_npu_available()
