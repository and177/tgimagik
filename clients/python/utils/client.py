import asyncio
import logging
import time
from text_generation import AsyncClient

from typing import List, Dict
from torch import IntTensor
from transformers import AutoTokenizer
import numpy as np

from hydra import (
    compose,
    initialize
)
from omegaconf import DictConfig, OmegaConf


log = logging.getLogger(__name__)
# intialize Hydra subsystem
initialize(version_base=None, config_path="conf")
cfg: DictConfig = compose("config.yaml")
print(f"Experiment configuration:\n{OmegaConf.to_yaml(cfg)}")

# serving system deployment
max_concurrent_requests: int = cfg.deployment.max_concurrent_requests
endpoint: str = ""
if cfg.deployment.local: endpoint = f"{cfg.deployment.addr}:{ cfg.deployment.port}"
if not cfg.deployment.local: endpoint = cfg.deployment.endpoint
client = AsyncClient(endpoint)

# uncomment for using natural language prompts
# prompt: str = "Hello" * 100
# output_lenght (decoding length)
max_new_tokens: int = cfg.GenerationConfig.max_new_tokens
# generation_strategy
repetition_penalty: float = cfg.GenerationConfig.repetition_penalty
do_sample: bool = cfg.GenerationConfig.do_sample
# generation finish reason
stop_sequences: List[str] = [stop for stop in cfg.GenerationConfig.stop_sequences]
tokenizer = AutoTokenizer.from_pretrained(cfg.macros.tokenizer_name)
# "id": 21820 = "Hello"
dummy_input: IntTensor = IntTensor([[21820]])
# inverse tokenization (sanity check)
# token: str = tokenizer.decode(dummy_input[0], skip_special_tokens=True)
sequence_length: int = cfg.GenerationConfig.sequence_length
multiple_dummy_repeat = dummy_input.repeat(1, sequence_length)
prompt: str = tokenizer.decode(multiple_dummy_repeat[0], skip_special_tokens=True)
assert np.shape(multiple_dummy_repeat)[-1] == sequence_length

generate_kwargs: Dict = {"do_sample": do_sample,
                         "max_new_tokens": max_new_tokens,
                        "repetition_penalty": repetition_penalty,
                        "stop_sequences": stop_sequences,
                        "decoder_input_details": True,
                        }

prompts: List = [prompt] * max_concurrent_requests
# create many coroutines
coros = [client.generate(prompt, **generate_kwargs) for prompt in prompts]

async def batch():
    return await asyncio.gather(*coros)
    
st: float = time.perf_counter_ns()
results = asyncio.run(batch())
et = (time.perf_counter_ns() - st) * 1e-9
print(f"Serving elapsed time: {et:0.4f} seconds")
# check the last response
print(results[-1].details)

total_input_sequence_tokens: int = 0
total_decoded_tokens: int = 0

for prompt, response in zip(prompts,results):
    # uncomment for see generations
    # print(prompt + response.generated_text)
    # assert np.shape(tokenizer(response.generated_text, return_tensors="pt").input_ids)[-1] -1 == max_new_tokens
    # assert response.details.generated_tokens == max_new_tokens
    total_input_sequence_tokens += len(response.details.prefill)
    total_decoded_tokens += response.details.generated_tokens
assert total_input_sequence_tokens == max_concurrent_requests * sequence_length

# stats
print(f"Serving elapsed time: {et:0.4f} seconds")
print(f"Total requests: {max_concurrent_requests}")
print(f"Total sequences tokens: {total_input_sequence_tokens}")
print(f"Sequence length: {sequence_length}")
print(f"Total decoded tokens: {total_decoded_tokens}")
print(f"Throughput: {total_decoded_tokens/et} tokens/sec")
print(f"Total processed tokens: {total_decoded_tokens + total_input_sequence_tokens}")