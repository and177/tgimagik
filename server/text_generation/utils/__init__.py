from text_generation.utils.convert import convert_file, convert_files
from text_generation.utils.dist import initialize_torch_distributed
from text_generation.utils.hub import (
    weight_files,
    weight_hub_files,
    download_weights,
    EntryNotFoundError,
    LocalEntryNotFoundError,
)
from text_generation.utils.tokens import (
    Greedy,
    NextTokenChooser,
    Sampling,
    StoppingCriteria,
)

__all__ = [
    "convert_file",
    "convert_files",
    "initialize_torch_distributed",
    "weight_files",
    "weight_hub_files",
    "download_weights",
    "EntryNotFoundError",
    "LocalEntryNotFoundError",
    "Greedy",
    "NextTokenChooser",
    "Sampling",
    "StoppingCriteria",
]
