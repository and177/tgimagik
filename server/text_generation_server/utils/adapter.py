import os
from collections import defaultdict
from pathlib import Path
from typing import List, Dict, Set, Tuple

import torch
from loguru import logger
from peft import LoraConfig
from peft.utils import transpose
from safetensors.torch import load_file, save_file
from tqdm import tqdm

from text_generation_server.utils.hub import weight_files


def compute_delta_weight(
    lora_A: torch.Tensor, 
    lora_B: torch.Tensor, 
    fan_in_fan_out: bool, 
    alpha: float, 
    r: float
) -> torch.Tensor:
    """Computes the delta weight for a Linear layer given A and B LoRA matrices.
    
    TODO: add logic for other module types beyond Linear layers.
    
    Reference: https://github.com/huggingface/peft/blob/v0.4.0/src/peft/tuners/lora.py#L799-L806
    """
    scaling = alpha / r
    delta_weight = transpose(lora_B @ lora_A, fan_in_fan_out) * scaling
    return delta_weight


def merge_adapter_weights(
    model_weights: Dict[str, torch.Tensor], 
    adapter_weights: Dict[str, torch.Tensor], 
    adapter_config: LoraConfig
) -> Tuple[Dict[str, torch.Tensor], Set[str]]:
    """Merges the adapter weights into the model weights."""
    module_mapping = defaultdict(dict)
    processed_adapter_weight_names = set()

    # map the original tensor names to their adapter counterparts
    for weight_name in model_weights:
        end_idx = weight_name.rfind(".weight")
        key = weight_name[:end_idx]
        for adapter_weight_name in adapter_weights:
            if key in adapter_weight_name:
                # example value: 'base_model.model.model.layers.10.self_attn.v_proj.lora_B.weight'
                # matrix_type gets the second to last element in the module name, i.e. 'lora_B'
                matrix_type = adapter_weight_name.split(".")[-2]
                module_mapping[weight_name][matrix_type] = adapter_weight_name
                processed_adapter_weight_names.add(adapter_weight_name)
    
    # merge adapter weights into model weights
    merged_weights = {}
    for weight_name, adapter_weight_names in tqdm(
        module_mapping.items(), desc="Merging adapter weights", total=len(module_mapping)):

        # TODO: support adapter types beyond LoRA
        lora_A = adapter_weights[adapter_weight_names["lora_A"]]
        lora_B = adapter_weights[adapter_weight_names["lora_B"]]
        delta_weight = compute_delta_weight(
            lora_A, lora_B, adapter_config.fan_in_fan_out, adapter_config.lora_alpha, adapter_config.r)
        merged_weights[weight_name] = model_weights[weight_name] + delta_weight
    return merged_weights, processed_adapter_weight_names


def create_merged_weight_files(
    adapter_id: str, 
    model_id: str,
    model_weight_filenames: List[Path]
) -> List[Path]:
    """Creates merged weight files for the given adapter ID and filenames."""
    adapter_filenames = weight_files(adapter_id, extension=".safetensors")
    
    adapter_config = LoraConfig.from_pretrained(adapter_id)
    if adapter_config.base_model_name_or_path != model_id:
        raise ValueError(f"Adapter {adapter_id} is not compatible with model {model_id}")
    
    # load adapter weights from all shards (should have relatively small memory footprint)
    adapter_weights = {}
    for filename in adapter_filenames:
        adapter_weights.update(load_file(filename))
    remaining_adapter_weight_names = set(adapter_weights.keys())

    merged_weight_directory = f"/data/{adapter_id.replace('/', '--')}-merged/"
    # just grab the existing files if they already exist and return immediately
    if os.path.exists(merged_weight_directory):
        logger.info("Merged weight files already exist, skipping merge computation.")
        return weight_files(merged_weight_directory)

    os.makedirs(merged_weight_directory)
    merged_weight_filenames = []
    for filename in model_weight_filenames:
        model_weights = load_file(filename)
        merged_weights, processed_adapter_weight_names = merge_adapter_weights(
            model_weights, adapter_weights, adapter_config)
        
        merged_adapter_filename = Path(merged_weight_directory, os.path.basename(filename))
        save_file(merged_weights, merged_adapter_filename)
        logger.debug(f"Saved merged weights into {merged_adapter_filename}")

        merged_weight_filenames.append(merged_adapter_filename)
        remaining_adapter_weight_names = remaining_adapter_weight_names.difference(
            processed_adapter_weight_names)
    
    if len(remaining_adapter_weight_names) > 0:
        logger.warning("WARNING: The following lora weights were not merged into the model weights:")
        for lora_name in remaining_adapter_weight_names:
            logger.warning("\t" + lora_name)

    return merged_weight_filenames


def main():
    adapter_id = "arnavgrg/codealpaca-qlora"
    adapter_config = LoraConfig.from_pretrained(adapter_id)
    model_id = adapter_config.base_model_name_or_path
    model_weight_filenames = weight_files(model_id, extension=".safetensors")
    
    merged_adapter_filenames = create_merged_weight_files(adapter_id, model_id, model_weight_filenames)
    print(merged_adapter_filenames)


if __name__ == '__main__':
    main()