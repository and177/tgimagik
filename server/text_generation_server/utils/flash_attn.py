import os
import torch

from loguru import logger
import math

from text_generation_server.utils.import_utils import (
    IS_CUDA_SYSTEM,
    IS_ROCM_SYSTEM,
    IS_XPU_SYSTEM,
)
from text_generation_server.utils.flash_attn_triton import triton_attention

if os.getenv("USE_FLASH_ATTENTION", "").lower() == "false":
    raise ImportError("`USE_FLASH_ATTENTION` is false.")

if not torch.cuda.is_available():
    raise ImportError("CUDA is not available")

major, minor = torch.cuda.get_device_capability()
is_sm75 = major == 7 and minor == 5
is_sm8x = major == 8 and minor >= 0
is_sm90 = major == 9 and minor == 0
is_sm94 = major == 9 and minor == 4

HAS_FLASH_ATTN = False
HAS_FLASH_ATTN_V2_CUDA = False
HAS_FLASH_ATTN_V2_ROCM = False

ROCM_USE_FLASH_ATTN_V2_CK = False
ROCM_USE_FLASH_ATTN_V2_TRITON = False

if IS_ROCM_SYSTEM:
    if os.getenv("ROCM_USE_FLASH_ATTN_V2_TRITON", "").lower() == "true":
        ROCM_USE_FLASH_ATTN_V2_TRITON = True
        logger.info("ROCm: using Flash Attention 2 Triton implementation.")
    else:
        ROCM_USE_FLASH_ATTN_V2_CK = True
        logger.info("ROCm: using Flash Attention 2 Composable Kernel implementation.")

try:
    try:
        import flash_attn_2_cuda
    except ImportError:
        architecture_suffix = ""
        if IS_CUDA_SYSTEM:
            architecture_suffix = "-cuda"
        elif IS_ROCM_SYSTEM:
            architecture_suffix = "-rocm"
        raise ImportError(
            "Flash Attention V2 is not installed.\n"
            "Use the official Docker image (ghcr.io/huggingface/text-generation-inference:latest) "
            f"or install flash attention v2 with `cd server && make install install-flash-attention-v2{architecture_suffix}`"
        )
    if IS_CUDA_SYSTEM and not (is_sm8x or is_sm90):
        raise ImportError(
            f"GPU with CUDA capability {major} {minor} is not supported for "
            "Flash Attention V2"
        )
    elif IS_ROCM_SYSTEM and not (is_sm8x or is_sm90 or is_sm94):
        raise ImportError(
            f"AMD GPU with compute capability {major} {minor} is not supported for "
            "Flash Attention V2"
        )
    HAS_FLASH_ATTN_V2_CUDA = IS_CUDA_SYSTEM
    HAS_FLASH_ATTN_V2_ROCM = IS_ROCM_SYSTEM
except ImportError as e:
    try:
        import flash_attn_cuda
    except ImportError:
        raise ImportError(
            "Flash Attention is not installed.\n"
            "Use the official Docker image (ghcr.io/huggingface/text-generation-inference:latest) "
            "or install flash attention with `cd server && make install install-flash-attention`"
        ) from e

    if IS_CUDA_SYSTEM and not (is_sm75 or is_sm8x or is_sm90):
        raise ImportError(
            f"GPU with CUDA capability {major} {minor} is not supported"
        ) from e
    elif IS_ROCM_SYSTEM:
        for idx in range(torch.cuda.device_count()):
            if "MI210" not in torch.cuda.get_device_name(
                idx
            ) and "MI250" not in torch.cuda.get_device_name(idx):
                raise ImportError(
                    f"AMD GPU {torch.cuda.get_device_name(idx)} does not support flash-attention"
                )

    logger.warning(f"Unable to use Flash Attention V2: {e}")
    HAS_FLASH_ATTN = True


def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
    """
    This is the equivalent of torch.repeat_interleave(x, dim=1, repeats=n_rep). The hidden states go from (batch,
    num_key_value_heads, seqlen, head_dim) to (batch, num_attention_heads, seqlen, head_dim)
    """
    total_tokens, num_key_value_heads, head_dim = hidden_states.shape
    if n_rep == 1:
        return hidden_states
    hidden_states = hidden_states[:, :, None, :].expand(
        total_tokens, num_key_value_heads, n_rep, head_dim
    )
    return hidden_states.reshape(total_tokens, num_key_value_heads * n_rep, head_dim)


def attention(
    q,
    k,
    v,
    out,
    cu_seqlens,
    max_s,
    softmax_scale,
    window_size_left=-1,
):
    if window_size_left <= 0 and window_size_left != -1:
        raise ValueError("`window_size_left` must be > 0 or -1")

    if IS_XPU_SYSTEM:
        if window_size_left != -1:
            raise ValueError(
                f"XPU version of Flash Attention does not support window attention (window_size_left != -1, got window_size_left={window_size_left})."
            )
        return ipex.llm.functional.varlen_attention(
            q,
            k,
            v,
            out,
            cu_seqlens,
            cu_seqlens,
            max_s,
            max_s,
            0.0,
            softmax_scale,
            False,
            True,
            False,
            None,
        )

    if IS_CUDA_SYSTEM and HAS_FLASH_ATTN_V2_CUDA:
        return flash_attn_2_cuda.varlen_fwd(
            q,
            k,
            v,
            out,
            cu_seqlens,
            cu_seqlens,
            None,
            None,
            None,
            max_s,
            max_s,
            0.0,
            softmax_scale,
            False,
            True,
            window_size_left,
            0,
            False,
            None,
        )
    elif IS_CUDA_SYSTEM and HAS_FLASH_ATTN:
        if window_size_left != -1:
            raise NotImplementedError(
                "window_size_left is only available with flash attn v2"
            )

        # Flash attention v1 requires q, k and v to have the same number of heads
        if k.shape[1] != q.shape[1]:
            # MQA expand
            if k.shape[1] == 1:
                k = k.expand(-1, q.shape[1], -1)
            # Grouped attention reshape
            else:
                original_shape = k.shape
                k = (
                    k.unsqueeze(2)
                    .expand(-1, -1, q.shape[1] // k.shape[1], -1)
                    .reshape(original_shape[0], -1, original_shape[2])
                )
        if v.shape[1] != q.shape[1]:
            # MQA expand
            if v.shape[1] == 1:
                v = v.expand(-1, q.shape[1], -1)
            # Grouped attention reshape
            else:
                original_shape = v.shape
                v = (
                    v.unsqueeze(2)
                    .expand(-1, -1, q.shape[1] // v.shape[1], -1)
                    .reshape(original_shape[0], -1, original_shape[2])
                )

        return flash_attn_cuda.fwd(
            q,
            k,
            v,
            out,
            cu_seqlens,
            cu_seqlens,
            max_s,
            max_s,
            0.0,
            softmax_scale,
            False,
            True,
            False,
            0,
            None,
        )
    elif IS_ROCM_SYSTEM and HAS_FLASH_ATTN_V2_ROCM and ROCM_USE_FLASH_ATTN_V2_CK:
        if window_size_left != -1:
            raise ValueError(
                f"RoCm version of Flash Attention v2 does not support window attention (window_size_left != -1, got window_size_left={window_size_left})."
            )

        # RoCm flash API does not take the window_size_left and window_size_right arguments.
        return flash_attn_2_cuda.varlen_fwd(
            q,
            k,
            v,
            out,
            cu_seqlens,
            cu_seqlens,
            max_s,
            max_s,
            0.0,
            softmax_scale,
            False,
            True,
            False,
            None,
        )
    elif IS_ROCM_SYSTEM and ROCM_USE_FLASH_ATTN_V2_TRITON:
        # NOTE: The Triton kernel silently outputs wrong results when using MQA/GQA and not
        # repeating.
        # TODO: just a sketch. Kind of need to abstract this `attention` function to enable some customization and pass those - let's sync with Nicolas for which implem he'd like
        
        # num_heads = q.shape[1]
        # num_kv_heads = k.shape[1]
        # if num_kv_heads != num_heads:
        #     # Interleave for MQA workaround.
        #     k = repeat_kv(k, num_heads // num_kv_heads)
        #     v = repeat_kv(v, num_heads // num_kv_heads)

        output, _ = triton_attention(
            q,
            k,
            v,
            out,
            cu_seqlens,
            cu_seqlens,
            max_s,
            max_s,
            True,
            softmax_scale,
        )
        return output
    else:
        raise NotImplementedError(
            f"Flash attention is not installed (IS_CUDA_SYSTEM={IS_CUDA_SYSTEM}, IS_ROCM_SYSTEM={IS_ROCM_SYSTEM}, HAS_FLASH_ATTN_V2_CUDA={HAS_FLASH_ATTN_V2_CUDA}, HAS_FLASH_ATTN_V2_ROCM={HAS_FLASH_ATTN_V2_ROCM})"
        )
