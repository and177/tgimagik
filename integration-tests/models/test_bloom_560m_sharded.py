import pytest

from utils import health_check


@pytest.fixture(scope="module")
def bloom_560m_sharded(launcher):
    with launcher("bigscience/bloom-560m", num_shard=2) as client:
        yield client


@pytest.mark.asyncio
async def test_bloom_560m_sharded(bloom_560m_sharded, snapshot):
    await health_check(bloom_560m_sharded, 60)

    response = await bloom_560m_sharded.generate("Test request", max_new_tokens=10)

    assert response.details.generated_tokens == 10
    assert response == snapshot


@pytest.mark.asyncio
async def test_bloom_560m_sharded_load(bloom_560m_sharded, generate_load, snapshot):
    await health_check(bloom_560m_sharded, 60)

    responses = await generate_load(
        bloom_560m_sharded, "Test request", max_new_tokens=10, n=4
    )

    assert len(responses) == 4

    assert responses == snapshot
