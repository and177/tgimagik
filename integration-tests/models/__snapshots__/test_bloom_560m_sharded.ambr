# serializer version: 1
# name: test_bloom_560m_sharded
  dict({
    'details': dict({
      'best_of_sequences': None,
      'finish_reason': <FinishReason.Length: 'length'>,
      'generated_tokens': 10,
      'prefill': list([
        dict({
          'id': 10264,
          'logprob': None,
          'text': 'Test',
        }),
        dict({
          'id': 8821,
          'logprob': -11.375,
          'text': ' request',
        }),
      ]),
      'seed': None,
      'tokens': list([
        dict({
          'id': 11,
          'logprob': -2.734375,
          'special': False,
          'text': '(',
        }),
        dict({
          'id': 5,
          'logprob': -1.9765625,
          'special': False,
          'text': '"',
        }),
        dict({
          'id': 1587,
          'logprob': -3.140625,
          'special': False,
          'text': 'get',
        }),
        dict({
          'id': 5,
          'logprob': -3.515625,
          'special': False,
          'text': '"',
        }),
        dict({
          'id': 15,
          'logprob': -0.37304688,
          'special': False,
          'text': ',',
        }),
        dict({
          'id': 567,
          'logprob': -2.6875,
          'special': False,
          'text': ' "',
        }),
        dict({
          'id': 17,
          'logprob': -4.65625,
          'special': False,
          'text': '.',
        }),
        dict({
          'id': 5,
          'logprob': -2.28125,
          'special': False,
          'text': '"',
        }),
        dict({
          'id': 12,
          'logprob': -1.6640625,
          'special': False,
          'text': ')',
        }),
        dict({
          'id': 30,
          'logprob': -0.14355469,
          'special': False,
          'text': ';',
        }),
      ]),
    }),
    'generated_text': '("get", ".");',
  })
# ---
# name: test_bloom_560m_sharded_load
  list([
    Response(generated_text='("get", ".");', details=Details(finish_reason=<FinishReason.Length: 'length'>, generated_tokens=10, seed=None, prefill=[PrefillToken(id=10264, text='Test', logprob=None), PrefillToken(id=8821, text=' request', logprob=-11.375)], tokens=[Token(id=11, text='(', logprob=-2.734375, special=False), Token(id=5, text='"', logprob=-2.328125, special=False), Token(id=1587, text='get', logprob=-2.953125, special=False), Token(id=5, text='"', logprob=-3.546875, special=False), Token(id=15, text=',', logprob=-0.359375, special=False), Token(id=567, text=' "', logprob=-2.484375, special=False), Token(id=17, text='.', logprob=-4.0625, special=False), Token(id=5, text='"', logprob=-2.46875, special=False), Token(id=12, text=')', logprob=-1.671875, special=False), Token(id=30, text=';', logprob=-0.18164062, special=False)], best_of_sequences=None)),
    Response(generated_text='("get", ".");', details=Details(finish_reason=<FinishReason.Length: 'length'>, generated_tokens=10, seed=None, prefill=[PrefillToken(id=10264, text='Test', logprob=None), PrefillToken(id=8821, text=' request', logprob=-11.375)], tokens=[Token(id=11, text='(', logprob=-2.734375, special=False), Token(id=5, text='"', logprob=-2.328125, special=False), Token(id=1587, text='get', logprob=-2.953125, special=False), Token(id=5, text='"', logprob=-3.546875, special=False), Token(id=15, text=',', logprob=-0.359375, special=False), Token(id=567, text=' "', logprob=-2.484375, special=False), Token(id=17, text='.', logprob=-4.0625, special=False), Token(id=5, text='"', logprob=-2.46875, special=False), Token(id=12, text=')', logprob=-1.671875, special=False), Token(id=30, text=';', logprob=-0.18164062, special=False)], best_of_sequences=None)),
    Response(generated_text='("get", ".");', details=Details(finish_reason=<FinishReason.Length: 'length'>, generated_tokens=10, seed=None, prefill=[PrefillToken(id=10264, text='Test', logprob=None), PrefillToken(id=8821, text=' request', logprob=-11.375)], tokens=[Token(id=11, text='(', logprob=-2.734375, special=False), Token(id=5, text='"', logprob=-2.328125, special=False), Token(id=1587, text='get', logprob=-2.953125, special=False), Token(id=5, text='"', logprob=-3.546875, special=False), Token(id=15, text=',', logprob=-0.359375, special=False), Token(id=567, text=' "', logprob=-2.484375, special=False), Token(id=17, text='.', logprob=-4.0625, special=False), Token(id=5, text='"', logprob=-2.46875, special=False), Token(id=12, text=')', logprob=-1.671875, special=False), Token(id=30, text=';', logprob=-0.18164062, special=False)], best_of_sequences=None)),
    Response(generated_text='("get", ".");', details=Details(finish_reason=<FinishReason.Length: 'length'>, generated_tokens=10, seed=None, prefill=[PrefillToken(id=10264, text='Test', logprob=None), PrefillToken(id=8821, text=' request', logprob=-11.375)], tokens=[Token(id=11, text='(', logprob=-2.734375, special=False), Token(id=5, text='"', logprob=-2.328125, special=False), Token(id=1587, text='get', logprob=-2.953125, special=False), Token(id=5, text='"', logprob=-3.546875, special=False), Token(id=15, text=',', logprob=-0.359375, special=False), Token(id=567, text=' "', logprob=-2.484375, special=False), Token(id=17, text='.', logprob=-4.0625, special=False), Token(id=5, text='"', logprob=-2.46875, special=False), Token(id=12, text=')', logprob=-1.671875, special=False), Token(id=30, text=';', logprob=-0.18164062, special=False)], best_of_sequences=None)),
  ])
# ---
